drivers/python/rethinkdb/net_twisted.py

```diff
@@ -0,0 +1,360 @@
+            if data >= 12:
+                token, length = struct.unpack('<qL', data[:12])
+                self.buf_token = token
+                self.buf_expected_length = length
+                data = data[12:]
+            else:
+                self.buf += data
+                # We quit the function, it is impossible to have read the
+                # entire payload at this point.
+                return
+
+        # 2. Buffer the data, until the size of the data match the expected
+        # length provided by the header.
+        self.buf += data
+        if len(self.buf) == self.buf_expected_length:
+            self.factory.response_handler(self.buf_token, self.buf)
+            self.buf_token = None
+            self.resetBuffer()
+
+    def dataReceived(self, data):
+        try:
+            self._handlers[self.state](data)
+        except Exception as e:
+            raise RqlDriverError('Driver failed to handle received data.'
+            'Error: {exc}. Dropping the connection.'.format(exc=str(e)))
+            self.transport.loseConnection()
+
+class DatabaseProtoFactory(ClientFactory):
+
+    protocol = DatabaseProtocol
+
+    def __init__(self, timeout, response_handler, handshake_payload):
+        self.timeout = timeout
+        self.handshake_payload = handshake_payload
+        self.response_handler = response_handler
+
+    def startedConnecting(self, connector):
+        pass
+
+    def buildProtocol(self, addr):
+        p = DatabaseProtocol(self)
+        return p
+
+    def clientConnectionLost(self, connector, reason):
+        pass
+
+    def clientConnectionFailed(self, connector, reason):
+        pass
+
+class CursorItems(DeferredQueue):
+
+    def __init__(self):
+        super(CursorItems, self).__init__(self)
+
+    def __len__(self):
+        return len(self.pending)
+
+class TwistedCursor(Cursor):
+
+    def __init__(self, *args, **kwargs):
+        super(TwistedCursor, self).__init__(*args, **kwargs)
+        self.new_response = Deferred()
+        self.items = CursorItems()
+
+    def _extend_queue(self, data):
+        for k in data:
+            self.items.put(k)
+
+    def _extend(self, res):
+        self.outstanding_requests -= 1
+        self.threshold = len(res.data)
+        if self.error is None:
+            if res.type == pResponse.SUCCESS_PARTIAL:
+                self._extend_queue(res.data)
+            elif res.type == pResponse.SUCCESS_SEQUENCE:
+                self._extend_queue(res.data)
+                self.error = self._empty_error()
+            else:
+                self.error = res.make_error(self.query)
+            self._maybe_fetch_batch()
+
+        if self.outstanding_requests == 0 and self.error is not None:
+            del self.conn._cursor_cache[res.token]
+
+        self.new_response.callback(None)
+        self.new_response = Deferred()
+
+    def _empty_error(self):
+        return RqlCursorEmpty(self.query.term)
+
+    def __iter__(self):
```

Comment:
`__iter__` is not `ìnlineCallbacks`-decorated.
Then, even if I raise a `StopIteration`, it will not be caught in the `unwindGenerator`.
All I do is to yield deferred.

This way, the user can do anything he wants.
Resolve those deferreds, reject them, pause them, cancel them.